### DISCUSSION TITLE
Is "robust code" supposed to avoid relying on scopeId or NestedNode Ids?

### DISCUSSION THREAD

**Discussion by xenia-foxtrot** [capnproto/capnproto#2127] [Q&A] [answered]
I'm currently writing some codegen for my C# cap'n'proto implementation. I'm building out the `CodeGeneratorRequest` from the schema into an object model, but I've run into some ambiguity in the schema's documentation:

The documentation for `scopeId` says:
> ID of the lexical parent node.  Typically, the scope node will have a NestedNode pointing back at this node, but robust code should avoid relying on this (and, in fact, group nodes are not listed in the outer struct's nestedNodes, since they are listed in the fields)

Generally, most people read this as "robust code should avoid relying on `NestedNode` pointing back at this node", but it's correct English for "this" to refer to `scopeId` as well.

This ambiguity is worsened by documentation for `NestedNode.id`:
> ID of the nested node.  Typically, the target node's scopeId points back to this node, but robust code should avoid relying on this.

Again, most people will read this as "robust code should avoid relying on the target node's `scopeId` pointing back to this node", which seems to contradict the documentation for `scopeId`. But, "this" could very well be referring to the nested node id as well.

So what's the correct reading? Should I rely on scopeId pointing to the parent, the NestedNode id pointing to the child, both, or neither?

**Comment by kentonv** [comment, marked as answer]
I think the point is that you should avoid writing code that assumes that these two pointers are consistent. E.g. if you write code that follows a `NestedNode`, it shouldn't crash if the `scopeId` of the nested node doesn't point back to the same parent.

For a concrete example, Cap'n Proto's SchemaLoader can load multiple versions of a schema (as long as they are compatible) and will expose the version determined to be newest. It could be that in one version, the type Foo was nested in Bar. In a later version, Foo was stand-alone and Bar didn't exist anymore. After loading both versions, you may find that you have a type definition for Bar which shows Foo as a nested node, but when you look at Foo's schema, you see the newer version whose `scopeId` doesn't point back at Bar.

Alternatively, it's possible that we'll add a notion of type aliases eventually where aliases show up as `NestedNode`s within the scope where the alias is defined, but the aliased type doesn't actually know anything about that scope.
